Intermittently seeing "sync: WaitGroup is reused before previous Wait has returned" in transport.go

I've seen the panic `"sync: WaitGroup is reused before the previous Wait has returned"` in a routine at `github.com/getsentry/sentry-go/transport.go:201`. Not sure how to debug this or how to explore the conditions. I can't find anything in the documentation that can help me explore the conditions post-panic. The stack trace leads me to...

```
vendor/github.com/getsentry/sentry-go/transport.go:201 +0x31 
github.com/getsentry/sentry-go.(*HTTPTransport).Flush 
vendor/github.com/getsentry/sentry-go/transport.go:200 +0x7c 
```

Any advice is appreciated on tools I can use to better break this open when it does occur!

Thank you @rhcarvalho for the explanation, I had the same issue, was doing flush per each report.
I will try what you suggest thanks :D 
```
panic: sync: WaitGroup is reused before previous Wait has returned
goroutine 52062 [running]:
sync.(*WaitGroup).Wait(0xc00006a458)
	/usr/local/go/src/sync/waitgroup.go:132 +0xae
github.com/getsentry/sentry-go.(*HTTPTransport).Flush.func1(0xc00006a420, 0xc0022013e0)
	/go/pkg/mod/github.com/getsentry/sentry-go@v0.3.0/transport.go:201 +0x31
created by github.com/getsentry/sentry-go.(*HTTPTransport).Flush
	/go/pkg/mod/github.com/getsentry/sentry-go@v0.3.0/transport.go:200 +0x7c
```

This test can be used to reproduce the issue:

```go
type testWriter testing.T

func (t *testWriter) Write(p []byte) (int, error) {
	t.Logf("%s", p)
	return len(p), nil
}

func TestHTTPTransportFlush(t *testing.T) {
	var counter uint64
	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		t.Logf("[SERVER] received event: #%d", atomic.AddUint64(&counter, 1))
	}))
	defer ts.Close()

	Logger.SetOutput((*testWriter)(t))

	tr := NewHTTPTransport()
	tr.Configure(ClientOptions{
		Dsn:        fmt.Sprintf("https://user@%s/42", ts.Listener.Addr()),
		HTTPClient: ts.Client(),
	})

	var wg sync.WaitGroup
	for i := 0; i < 2; i++ {
		i := i
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 2; j++ {
				t.Logf("tr.SendEvent #%d from goroutine #%d", j, i)
				tr.SendEvent(NewEvent())
				ok := tr.Flush(100 * time.Millisecond)
				if !ok {
					t.Error("Flush() timed out")
				}
			}
		}()
	}
	wg.Wait()
}
```

Example run:

```
$ go test -failfast -run TestHTTPTransportFlush -count=100 -v
=== RUN   TestHTTPTransportFlush
--- PASS: TestHTTPTransportFlush (0.00s)
    transport_test.go:189: tr.SendEvent #0 from goroutine #1
    transport_test.go:189: tr.SendEvent #0 from goroutine #0
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Sending  event [] to 127.0.0.1 project: 42
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Sending  event [] to 127.0.0.1 project: 42
    transport_test.go:170: [SERVER] received event: #1
    transport_test.go:170: [SERVER] received event: #2
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Buffer flushed successfully.
    transport_test.go:189: tr.SendEvent #1 from goroutine #0
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Buffer flushed successfully.
    transport_test.go:189: tr.SendEvent #1 from goroutine #1
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Sending  event [] to 127.0.0.1 project: 42
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Sending  event [] to 127.0.0.1 project: 42
    transport_test.go:170: [SERVER] received event: #3
    transport_test.go:170: [SERVER] received event: #4
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Buffer flushed successfully.
    transport_test.go:163: [Sentry] 2020/01/14 22:07:51 Buffer flushed successfully.
=== RUN   TestHTTPTransportFlush
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 15 [running]:
sync.(*WaitGroup).Wait(0xc0000204b8)
        /usr/lib/golang/src/sync/waitgroup.go:132 +0xad
github.com/getsentry/sentry-go.(*HTTPTransport).Flush.func1(0xc000020480, 0xc00002c8a0)
        .../sentry-go/transport.go:205 +0x31
created by github.com/getsentry/sentry-go.(*HTTPTransport).Flush
        .../sentry-go/transport.go:204 +0x7c
exit status 2
FAIL    github.com/getsentry/sentry-go  0.012s
```

Without instrumenting `HTTPTransport.Flush` to enforce an specific order of events, the test is flaky by nature. What's more, the panic is caused by an implementation detail. Since I'd rather be testing behavior, and I certainly do not want flaky tests, I am inclined not to add such test together with the fix.

Posting it here for transparency and future reference.

--- a/transport.go
+++ b/transport.go
@@ -96,16 +96,25 @@ func getRequestBodyFromEvent(event *Event) []byte {
 // HTTPTransport
 // ================================
 
+// A batch groups items that are processed sequentially.
+type batch struct {
+	items   chan *http.Request
+	started chan struct{} // closed to signal items started to be worked on
+	done    chan struct{} // closed to signal completion of all items
+}
+
 // HTTPTransport is a default implementation of `Transport` interface used by `Client`.
 type HTTPTransport struct {
 	dsn       *Dsn
 	client    *http.Client
 	transport *http.Transport
 
-	buffer        chan *http.Request
+	// buffer is a channel of batches. Calling Flush terminates work on the
+	// current in-flight items and starts a new batch for subsequent events.
+	buffer chan batch
+
 	disabledUntil time.Time
 
-	wg    sync.WaitGroup
 	start sync.Once
 
 	// Size of the transport buffer. Defaults to 30.
@@ -130,9 +139,17 @@ func (t *HTTPTransport) Configure(options ClientOptions) {
 		Logger.Printf("%v\n", err)
 		return
 	}
-
 	t.dsn = dsn
-	t.buffer = make(chan *http.Request, t.BufferSize)
+
+	// A buffered channel with capacity 1 works like a mutex, ensuring only one
+	// goroutine can access the current batch at a given time. Access is
+	// synchronized by reading from and writing to the channel.
+	t.buffer = make(chan batch, 1)
+	t.buffer <- batch{
+		items:   make(chan *http.Request, t.BufferSize),
+		started: make(chan struct{}),
+		done:    make(chan struct{}),
+	}
 
 	if options.HTTPTransport != nil {
 		t.transport = options.HTTPTransport
@@ -178,10 +195,21 @@ func (t *HTTPTransport) SendEvent(event *Event) {
 		request.Header.Set(headerKey, headerValue)
 	}
 
-	t.wg.Add(1)
+	// <-t.buffer is equivalent to acquiring a lock to access the current batch.
+	// A few lines below, t.buffer <- b releases the lock.
+	//
+	// The lock must be held during the select block below to guarantee that
+	// b.items is not closed while trying to send to it. Remember that sending
+	// on a closed channel panics.
+	//
+	// Note that the select block takes a bounded amount of CPU time because of
+	// the default case that is executed if sending on b.items would block. That
+	// is, the event is dropped if it cannot be sent immediately to the b.items
+	// channel (used as a queue).
+	b := <-t.buffer
 
 	select {
-	case t.buffer <- request:
+	case b.items <- request:
 		Logger.Printf(
 			"Sending %s event [%s] to %s project: %d\n",
 			event.Level,
@@ -190,51 +218,92 @@ func (t *HTTPTransport) SendEvent(event *Event) {
 			t.dsn.projectID,
 		)
 	default:
-		t.wg.Done()
 		Logger.Println("Event dropped due to transport buffer being full.")
-		// worker would block, drop the packet
 	}
+
+	t.buffer <- b
 }
 
 // Flush notifies when all the buffered events have been sent by returning `true`
 // or `false` if timeout was reached.
 func (t *HTTPTransport) Flush(timeout time.Duration) bool {
-	c := make(chan struct{})
+	toolate := time.After(timeout)
 
-	go func() {
-		t.wg.Wait()
-		close(c)
-	}()
+	// Wait until processing the current batch has started or the timeout.
+	//
+	// We must wait until the worker has seen the current batch, because it is
+	// the only way b.done will be closed. If we do not wait, there is a
+	// possible execution flow in which b.done is never closed, and the only way
+	// out of Flush would be waiting for the timeout, which is undesired.
+	var b batch
+	for {
+		select {
+		case b = <-t.buffer:
+			select {
+			case <-b.started:
+				goto started
+			default:
+				t.buffer <- b
+			}
+		case <-toolate:
+			goto fail
+		}
+	}
 
+started:
+	// Signal that there won't be any more items in this batch, so that the
+	// worker inner loop can end.
+	close(b.items)
+	// Start a new batch for subsequent events.
+	t.buffer <- batch{
+		items:   make(chan *http.Request, t.BufferSize),
+		started: make(chan struct{}),
+		done:    make(chan struct{}),
+	}
+
+	// Wait until the current batch is done or the timeout.
 	select {
-	case <-c:
+	case <-b.done:
 		Logger.Println("Buffer flushed successfully.")
 		return true
-	case <-time.After(timeout):
-		Logger.Println("Buffer flushing reached the timeout.")
-		return false
+	case <-toolate:
+		goto fail
 	}
+
+fail:
+	Logger.Println("Buffer flushing reached the timeout.")
+	return false
 }
 
 func (t *HTTPTransport) worker() {
-	for request := range t.buffer {
-		if time.Now().Before(t.disabledUntil) {
-			t.wg.Done()
-			continue
+	for b := range t.buffer {
+		// Signal that processing of the current batch has started.
+		close(b.started)
+
+		// Return the batch to the buffer so that other goroutines can use it.
+		// Equivalent to releasing a lock.
+		t.buffer <- b
+
+		// Process all batch items.
+		for request := range b.items {
+			if time.Now().Before(t.disabledUntil) {
+				continue
+			}
+
+			response, err := t.client.Do(request)
+
+			if err != nil {
+				Logger.Printf("There was an issue with sending an event: %v", err)
+			}
+
+			if response != nil && response.StatusCode == http.StatusTooManyRequests {
+				t.disabledUntil = time.Now().Add(retryAfter(time.Now(), response))
+				Logger.Printf("Too many requests, backing off till: %s\n", t.disabledUntil)
+			}
 		}
 
-		response, err := t.client.Do(request)
-
-		if err != nil {
-			Logger.Printf("There was an issue with sending an event: %v", err)
-		}
-
-		if response != nil && response.StatusCode == http.StatusTooManyRequests {
-			t.disabledUntil = time.Now().Add(retryAfter(time.Now(), response))
-			Logger.Printf("Too many requests, backing off till: %s\n", t.disabledUntil)
-		}
-
-		t.wg.Done()
+		// Signal that processing of the batch is done.
+		close(b.done)
 	}
 }
 

--- a/transport_test.go
+++ b/transport_test.go
@@ -1,7 +1,12 @@
 package sentry
 
 import (
+	"encoding/json"
+	"fmt"
 	"net/http"
+	"net/http/httptest"
+	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 )
@@ -151,4 +156,137 @@ func TestRetryAfterDateHeader(t *testing.T) {
 		},
 	}
 	assertEqual(t, retryAfter(now, &r), time.Second*13)
+}
+
+// A testHTTPServer counts events sent to it. It requires a call to Unblock
+// before incrementing its internal counter and sending a response to the HTTP
+// client. This allows for coordinating the execution flow when needed.
+type testHTTPServer struct {
+	*httptest.Server
+	// eventCounter counts the number of events processed by the server.
+	eventCounter *uint64
+	// ch is used to block/unblock the server on demand.
+	ch chan bool
+}
+
+func newTestHTTPServer(t *testing.T) *testHTTPServer {
+	ch := make(chan bool)
+	eventCounter := new(uint64)
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		var event struct {
+			EventID string `json:"event_id"`
+		}
+		dec := json.NewDecoder(r.Body)
+		err := dec.Decode(&event)
+		if err != nil {
+			t.Fatal(err)
+		}
+		// Block until signal to continue.
+		<-ch
+		count := atomic.AddUint64(eventCounter, 1)
+		t.Logf("[SERVER] {%.4s} event received (#%d)", event.EventID, count)
+	}
+	return &testHTTPServer{
+		Server:       httptest.NewTLSServer(http.HandlerFunc(handler)),
+		eventCounter: eventCounter,
+		ch:           ch,
+	}
+}
+
+func (ts *testHTTPServer) EventCount() uint64 {
+	return atomic.LoadUint64(ts.eventCounter)
+}
+
+func (ts *testHTTPServer) Unblock() {
+	ts.ch <- true
+}
+
+func TestHTTPTransport(t *testing.T) {
+	server := newTestHTTPServer(t)
+	defer server.Close()
+
+	transport := NewHTTPTransport()
+	transport.Configure(ClientOptions{
+		Dsn:        fmt.Sprintf("https://test@%s/1", server.Listener.Addr()),
+		HTTPClient: server.Client(),
+	})
+
+	// Helpers
+
+	transportSendTestEvent := func(t *testing.T) (id string) {
+		t.Helper()
+
+		e := NewEvent()
+		id = uuid()
+		e.EventID = EventID(id)
+
+		transport.SendEvent(e)
+		t.Logf("[CLIENT] {%.4s} event sent", e.EventID)
+		return id
+	}
+
+	transportMustFlush := func(t *testing.T, id string) {
+		t.Helper()
+
+		ok := transport.Flush(100 * time.Millisecond)
+		if !ok {
+			t.Fatalf("[CLIENT] {%.4s} Flush() timed out", id)
+		}
+	}
+
+	serverEventCountMustBe := func(t *testing.T, n uint64) {
+		t.Helper()
+
+		count := server.EventCount()
+		if count != n {
+			t.Fatalf("[SERVER] event count = %d, want %d", count, n)
+		}
+	}
+
+	// Actual tests
+
+	testSendSingleEvent := func(t *testing.T) {
+		// Sending a single event should increase the server event count by
+		// exactly one.
+
+		initialCount := server.EventCount()
+		id := transportSendTestEvent(t)
+
+		// Server is blocked waiting for us, right now count must not have
+		// changed yet.
+		serverEventCountMustBe(t, initialCount)
+
+		// After unblocking the server, Flush must guarantee that the server
+		// event count increased by one.
+		server.Unblock()
+		transportMustFlush(t, id)
+		serverEventCountMustBe(t, initialCount+1)
+	}
+	t.Run("SendSingleEvent", testSendSingleEvent)
+
+	t.Run("FlushMultipleTimes", func(t *testing.T) {
+		// Flushing multiple times should not increase the server event count.
+
+		initialCount := server.EventCount()
+		for i := 0; i < 10; i++ {
+			transportMustFlush(t, fmt.Sprintf("loop%d", i))
+		}
+		serverEventCountMustBe(t, initialCount)
+	})
+
+	t.Run("ConcurrentSendAndFlush", func(t *testing.T) {
+		// It should be safe to send events and flush concurrently.
+
+		var wg sync.WaitGroup
+		wg.Add(2)
+		go func() {
+			testSendSingleEvent(t)
+			wg.Done()
+		}()
+		go func() {
+			transportMustFlush(t, "from goroutine")
+			wg.Done()
+		}()
+		wg.Wait()
+	})
 }

